Lib/unittest/mock.py

```diff
@@ -2758,13 +2758,15 @@ def create_autospec(spec, spec_set=False, instance=False, _parent=None,
                                f'[object={spec!r}]')
     is_async_func = _is_async_func(spec)
 
-    entries = [(entry, _missing) for entry in dir(spec)]
+    base_entries = {entry: _missing for entry in dir(spec)}
     if is_type and instance and is_dataclass(spec):
         dataclass_fields = fields(spec)
```

Comment:
This is obscure enough that I think it's worth adding a comment explaining why the code works as it does:

```suggestion
        # Dataclass instance mocks created from a class may not have all of their fields
        # prepopulated with default values. Create an initial set of attribute entries from
        # the dataclass field annotations, but override them with the actual attribute types
        # when fields have already been populated.
        dataclass_fields = fields(spec)
```